### ISSUE TITLE
For nearest-neighbor remapping, ensure results are independent of processor count if there are equidistant source points

### ISSUE THREAD

**Issue by billsacks** [esmf-org/esmf#276]
For nearest-neighbor remapping, if there are equidistant source points, there is currently some logic that says that, if there are equidistant source points, arbitrarily use the point with the smallest ID. But, according to @oehmke , that logic isn't done in the multi-processor case, because currently the IDs aren't sent between processors. This results in nearest-neighbor mapping giving different results with different processor counts if there are equidistant source points. @oehmke proposes adding a send of the IDs so that the multi-processor case can break ties using the ID, similarly to in the single-processor case.

### Discussed in https://github.com/orgs/esmf-org/discussions/261

<div type='discussions-op-text'>

<sup>Originally posted by **samsrabin** July 10, 2024</sup>
### Requirements

- [X] Reviewed [ESMF Reference Manual](https://earthsystemmodeling.org/doc/)
- [X] Searched [GitHub Discussions](https://github.com/orgs/esmf-org/discussions?discussions_q=)

### Affiliation(s)

NSF-NCAR

### Issue

In [CTSM](https://github.com/ESCOMP/CTSM), we use ESMF to read some input files. One particular pair of input files, specifying crop sowing window start and end dates, is at half-degree resolution. We tell ESMF to do nearest-neighbor[^1] spatial interpolation as necessary to match the simulation grid.

[^1]: It needs to be nearest-neighbor because dates are modulo—interpolating between Jan. 2 [day 2] and Dec. 31 (day 365) should give Jan. 1 (day 1), not July 3-4 (day [2+365]/2 = 183.5)—and that's not something ESMF can do, to my knowledge.

When I do a run at 10°x15° resolution, some of the simulation gridcell centers are located exactly at the "corners" of four half-degree input pixels, meaning that those four neighbors are equally near. It doesn't matter to me which of those ESMF chooses as the "nearest neighbor," as long as it's consistent.

Unfortunately, it's not: At least one gridcell has a different "nearest neighbor" chosen _depending on how many processors the job is split across_.

As an example, I've made a figure based on two cases that are identical in setup except that Case 1 used 128 processors and Case 2 used 64. Due to this issue, a certain crop in the gridcell centered at latitude 0, longitude 30°E[^2] gets sowing window of days 7-82 in Case 1 and 336-46 in Case 2.

[^2]: There are other crops in this gridcell that also get different sowing windows. There are no crops in any other gridcell that get different sowing windows, but that doesn't necessarily mean different "nearest" neighbors are getting chosen. That might be happening, just with input pixels that don't differ.

The white/gray/black in this figure represents the half-degree sowing window files. Gray pixels match the values in Case 1, black pixels match Case 2, and white pixels match neither. The red lines intersect at the center of the 10x15 CTSM gridcell.
![screenshot_1104](https://github.com/esmf-org/esmf/assets/10454527/74af317f-2eaa-4c0d-bffe-78e46c4a08c8)
It looks like Case 1 reads from the pixel to the southwest, whereas Case 2 reads from the pixel to the northwest.

Some notes:
- I'm not 100% certain this is an ESMF issue as opposed to something weird that CTSM is doing, but I'm at the point where I've done all the troubleshooting I can within CTSM.
- This reproduces every time, over dozens of tests.

Tagging @ekluzek, @billsacks, and @briandobbins, who have expressed interest in this. By the way, I think I mentioned to y'all that I was having an ERP test pass but the equivalent PEM test fail—this is why! The read of sowing windows only happens at the very beginning of the test, so changing processor count halfway through makes no difference.

### Autotag

@oehmke</div>

**Comment by samsrabin** [comment]
Following up: Is this something that's on the roadmap to be in the ESMF version used in the CESM3 release? No worries if not, but in that case I'll need to make some of my tooling more robust and official.

**Comment by oehmke** [comment]
Yep, it's on the roadmap to ESMF 8.8.0, which is what we're targeting for CESM3. I'm hoping to get it done soon-ish, so we can make sure that it works awhile before the release.

**Comment by samsrabin** [comment]
Excellent, thanks!

**Comment by anntsay** [comment]
CESM person that needs this may have another workaround.. so punting to 8.9.0 for now.

**Comment by billsacks** [comment]
Ann's comment came from my brief/vague verbal comment. Yes, @samsrabin , I was referring to you here. My understanding was that you have a workaround that can/will be applied in the upcoming CESM3 release code, so the lack of a fix in ESMF won't be a hold-up for CESM3... but actually, I just read back through the comments here and see that we had given the message that the more robust fix would make it into 8.8, which now no longer looks like it will be possible. Sorry: that earlier comment came before we pushed the 8.8 release timing earlier by a couple of months.

So, @samsrabin , can you let us know how much of a problem it will be for you that the ESMF release that will be used in CESM3 likely won't have this more robust fix in place?

**Comment by samsrabin** [comment]
Very little problem at all, I just need to make the scripts I used for the tweaking more robust. Thanks!

**Comment by billsacks** [comment]
Thanks @samsrabin , and sorry about this. There were different forces pulling in different directions regarding the ESMF 8.8 release timing. It's still *possible* that ESMF 8.9 will be ready in time for CESM3, but that will depend largely on the CESM3 timing, so at this point we can't count on it.

**Comment by oehmke** [comment]
@billsacks or @samsrabin , this turned out to not be the id issue, but I think that I have figure out what it was. I have a branch (fix_nn) with a fix on it. Would it be hard to run your case on that branch to see if it works?  I'd be happy to build the branch on Derecho next week if that would make it easier.  Let me know. Thanks!

**Comment by samsrabin** [comment]
Great to hear, @oehmke! Yes, I would need to test this on Derecho, but I'd need guidance (@billsacks?) on how to set up CTSM to use your branch's ESMF.

**Comment by oehmke** [comment]
@samsrabin and @billsacks here is the path to the esmf.mk file:
/glade/work/oehmke/ESMF/fix_nn/lib/libO/Unicos.intel.64.mpi.default/esmf.mk

 Let me know if you need to me to do something else (built with another set of compilers).  Thanks!

**Comment by samsrabin** [comment]
Unfortunately it seems like we'll require updates to our codebase to make this work. My tests all failed build with errors like this in cesm.bldlog:
```
/glade/work/samrabin/ctsm_esmf-nn-bugfix/src/fates/biogeophys/FatesPlantHydraulicsMod.F90:5996: undefined reference to `dgesv_'
ld: /glade/derecho/scratch/samrabin/tests_0418-173427de/ERP_D_P128x1_Ld26.f10_f10_mg37.IHistClm60BgcCrop.derecho_intel.clm-RxCropCalsAdaptGGCMI.0418-173427de/bld/intel/mpich/debug/threads/lib//libclm.a(SoilWaterMovementMod.o): in function `soilwatermovementmod_mp_soilwater_moisture_form_':
/glade/work/samrabin/ctsm_esmf-nn-bugfix/src/biogeophys/SoilWaterMovementMod.F90:1281: undefined reference to `dgtsv_'
ld: /glade/derecho/scratch/samrabin/tests_0418-173427de/ERP_D_P128x1_Ld26.f10_f10_mg37.IHistClm60BgcCrop.derecho_intel.clm-RxCropCalsAdaptGGCMI.0418-173427de/bld/intel/mpich/debug/threads/lib//libclm.a(TwoStreamMLPEMod.o): in function `twostreammlpemod_mp_solve_':
/glade/work/samrabin/ctsm_esmf-nn-bugfix/src/fates/radiation/TwoStreamMLPEMod.F90:1580: undefined reference to `dgesv_'
ld: /glade/derecho/scratch/samrabin/tests_0418-173427de/ERP_D_P128x1_Ld26.f10_f10_mg37.IHistClm60BgcCrop.derecho_intel.clm-RxCropCalsAdaptGGCMI.0418-173427de/bld/intel/mpich/debug/threads/lib//libclm.a(UrbBuildTempOleson2015Mod.o): in function `urbbuildtempoleson2015mod_mp_buildingtemperature_':
/glade/work/samrabin/ctsm_esmf-nn-bugfix/src/biogeophys/UrbBuildTempOleson2015Mod.F90:663: undefined reference to `dgesv_'
ld: /glade/derecho/scratch/samrabin/tests_0418-173427de/ERP_D_P128x1_Ld26.f10_f10_mg37.IHistClm60BgcCrop.derecho_intel.clm-RxCropCalsAdaptGGCMI.0418-173427de/bld/intel/mpich/debug/threads/lib//libclm.a(BandDiagonalMod.o): in function `banddiagonalmod_mp_banddiagonal_':
/glade/work/samrabin/ctsm_esmf-nn-bugfix/src/biogeophys/BandDiagonalMod.F90:197: undefined reference to `dgbsv_'
gmake: *** [/glade/derecho/scratch/samrabin/tests_0418-173427de/ERP_D_P128x1_Ld26.f10_f10_mg37.IHistClm60BgcCrop.derecho_intel.clm-RxCropCalsAdaptGGCMI.0418-173427de/Tools/Makefile:935: ../../cesm.exe] Error 1
```

Hopefully it'll be as simple as changing function names?

Test list run directory is `/glade/derecho/scratch/samrabin/tests_0418-173427de/` if you want to look for yourself, with source directory `/glade/work/samrabin/ctsm_esmf-nn-bugfix/`.

**Comment by briandobbins** [comment]
Those are all LAPACK calls; my guess is the ESMF Bob built used the internal LAPACK but somehow is not including them in the  linking libraries?  Switching to the system LAPACK should address this, with no change inside CTSM, since that ensures the link flags get added (the MKL, in this case).

I'll take a look at building another version in a bit, Sam, so you can at least try that out.

**Comment by briandobbins** [comment]
I didn't get to try this yet, and need to hop on a call, but here's a new ESMF mk file you can try, with a system MKL / LAPACK build.

/glade/work/bdobbins/esmf_for_sam/fix_nn/installed/lib/esmf.mk

I'll give it a shot after my call too.

**Comment by briandobbins** [comment]
I just did a single test (one of yours that had failed), and it worked for me.  Give it a shot, and let me know if you run into any trouble.  Mine was here:

/glade/derecho/scratch/bdobbins/PEM_D_P128x1_Ld26.f10_f10_mg37.IHistClm60BgcCrop.derecho_intel.clm-RxCropCalsNoAdapt.20250419_130353_1pkjnc

Thanks!
  - Brian

**Comment by samsrabin** [comment]
Excellent, it looks like that branch solves the problem! Thanks so much, @oehmke. Any ideas when this will come into CESM?

**Comment by billsacks** [comment]
Great, thank you for all of your work on this – @oehmke for the fix, @briandobbins for the build and @samsrabin for testing and confirming that this fixes the issue!

This will come into the ESMF 8.9 release, which is scheduled for sometime this summer. So it will probably still be a few months before this can come into CESM. In the past, we have sometimes pointed to ESMF beta tags in CESM, but I think that's risky given that we're potentially getting close to code freeze for CESM, and the tentative plan has been to not require anything beyond ESMF 8.8 for the CESM3 release. So my feeling is that, as long as we want to stick with ESMF 8.8 as the ESMF version required for CESM3, then we need to keep in place the workaround for now and wait to bring in this update. @samsrabin and others - does this feel okay?

**Comment by samsrabin** [comment]
Yep! I think that's what we discussed before, but I just wanted to confirm. Thanks again.

**Comment by oehmke** [comment]
@briandobbins thanks for the fix for the linking issue.  We'll have to investigate why that happened. 

@samsrabin and @billsacks thanks for testing that. Good to hear that it fixes the issue. I'll merge it into develop soon, so that it'll be in the next release.
